File: app/src/main/java/com/monk/rxjava2/imetate2/Observable.java
package com.monk.rxjava2.imetate2;

/**
 * @author monk
 * @date 2019-03-01
 */
public abstract class Observable<T> {

    public static <T> Observable<T> create(ObservableOnSubscribe<T> observableOnSubscribe) {
        return new ObservableCreator<>(observableOnSubscribe);
    }

    public void subscribe(Observer<T> observer) {
        subscribeActual(observer);
    }

    public abstract void subscribeActual(Observer<T> observer);

}

File: module_fragment/src/main/java/com/monk/modulefragment/rxjava2/imetate2/Observable.kt
package com.monk.modulefragment.rxjava2.imetate2

/**
 * @author monk
 * @date 2019-03-01
 */
abstract class Observable<T> {
    fun subscribe(observer: Observer<T>?) {
        subscribeActual(observer)
    }

    abstract fun subscribeActual(observer: Observer<T>?)

    companion object {
        fun <T> create(observableOnSubscribe: ObservableOnSubscribe<T>?): Observable<T> {
            return ObservableCreator(observableOnSubscribe!!)
        }
    }
}